node/actors/network/src/gossip/attestation/mod.rs

```diff
@@ -0,0 +1,330 @@
+        let state = state.borrow();
+        let Some(state) = &*state else { return vec![] };
+        if &state.config.batch_to_attest != want {
+            return vec![];
+        }
+        state.votes.values().cloned().collect()
+    }
+
+    /// Waits for the certificate for a batch with the given number to be collected.
+    /// Returns None iff attestation already skipped to collecting certificate for some later batch.
+    pub async fn wait_for_qc(
+        &self,
+        ctx: &ctx::Ctx,
+        n: attester::BatchNumber,
+    ) -> ctx::OrCanceled<Option<attester::BatchQC>> {
+        let recv = &mut self.state.subscribe();
+        recv.mark_changed();
+        loop {
+            let state = sync::changed(ctx, recv).await?;
+            let Some(state) = state.as_ref() else {
+                continue;
+            };
+            if state.config.batch_to_attest.number < n {
+                continue;
+            };
+            if state.config.batch_to_attest.number > n {
+                return Ok(None);
+            }
+            if let Some(qc) = state.qc() {
+                return Ok(Some(qc));
+            }
+        }
+    }
+
+    /// Updates the attestation config.
+    /// Clears the votes collected for the previous config.
+    /// Batch number has to increase with each update.
+    pub async fn update_config(&self, config: Arc<Config>) -> anyhow::Result<()> {
+        let locked = self.state.lock().await;
+        let old = locked.borrow().clone();
+        if let Some(old) = old.as_ref() {
+            if *old.config == *config {
+                return Ok(());
+            }
+            anyhow::ensure!(
+                old.config.batch_to_attest.genesis == config.batch_to_attest.genesis,
+                "tried to change genesis"
+            );
+            anyhow::ensure!(
+                old.config.batch_to_attest.number < config.batch_to_attest.number,
+                "tried to decrease batch number"
+            );
+        }
+        let mut new = State {
+            config,
+            votes: im::HashMap::new(),
+            weight: 0,
+        };
+        if let Some(key) = self.key.as_ref() {
+            if new.config.committee.contains(&key.public()) {
+                let vote = key.sign_msg(new.config.batch_to_attest.clone());
+                // This is our own vote, so it always should be valid.
+                new.insert_vote(Arc::new(vote)).unwrap();
+            }
+        }
+        metrics::METRICS
+            .batch_number
+            .set(new.config.batch_to_attest.number.0);
+        metrics::METRICS
+            .committee_size
+            .set(new.config.committee.len());
+        metrics::METRICS.votes_collected.set(new.votes.len());
+        #[allow(clippy::float_arithmetic)]
+        metrics::METRICS
+            .weight_collected
+            .set(new.weight as f64 / new.config.committee.total_weight() as f64);
+        locked.send_replace(Some(new));
+        Ok(())
```

Comment:
let's add comprehensive logging to this. Metrics are good, but they are only scrapped every 30 seconds, we may go through multiple batches during that time - that is, we won't know if we skipped something

Reply:
done